rpc/src/rpc.rs

```diff
@@ -156,6 +157,7 @@ impl EthGetBlockByNumberResponse {
     async fn fetch_previous_block_hashes<U: IntoUrl + Copy>(
         rpc_url: U,
         block_number: u64,
+        block_hash_cache: &mut HashMap<u64, H256>,
     ) -> Result<Vec<H256>> {
```

Comment:
Quid of having this run outside of `fetch_prover_input`? As we're still calling `process_block ` on every individual block of the range, it feels a bit heavy, while block hash fetching could be done only once for the entire range of blocks passed as input, without the need for this caching `HashMap`, if it was performed over the entire range at once in `jerigon_main`.